libs/geometry/src/TangentSpaceMesh.cpp

```diff
@@ -255,16 +353,47 @@ Builder& Builder::algorithm(Algorithm algo) noexcept {
 }
 
 TangentSpaceMesh* Builder::build() {
-    mMesh->mOutput->algorithm = selectAlgorithm(mMesh->mInput);
-    // "Working in progress. Not for use"
-    return NULL;
+    // Work in progress. Not for use.
+    Algorithm algo = selectAlgorithm(mMesh->mInput);
+    MethodPtr method = nullptr;
+    switch (algo) {
+        case Algorithm::FRISVAD:
+            method = [](const TangentSpaceMeshInput* input, TangentSpaceMeshOutput* output) {
+                normalsOnlyMethod(input, output, frisvadKernel);
+            };
+            break;
+        case Algorithm::HUGHES_MOLLER:
+            method = [](const TangentSpaceMeshInput* input, TangentSpaceMeshOutput* output) {
+                normalsOnlyMethod(input, output, hughesMollerKernel);
+            };
+            break;
+        case Algorithm::FLAT_SHADING:
+            method = flatShadingMethod;
+            break;
+        default:
+            break;
+    }
+    assert_invariant(method);
+    method(mMesh->mInput, mMesh->mOutput);
+    return mMesh;
 }
 
 TangentSpaceMesh::TangentSpaceMesh() noexcept
         :mInput(new TangentSpaceMeshInput()), mOutput(new TangentSpaceMeshOutput()) {
```

Comment:
do these need to be on the heap? Why not just make them members? Also I don't see where they're `delete`'ed?

Reply:
Sorry, do you mean `mInput` `mOutput` ? We clear out any allocation in the destructor TangentSpaceMesh.  It's the client's responsibility to call delete on the mesh once it's computed.

Reply:
Actually I see what you're saying.  On the heap because the `.h` only sees the forward declaration of these two types, and the actual declaration is in the anonymous namespace here.  

I've delete them in the destructor.